File: CliAdapter/GuiProvider.cpp

#include <string>
#include <functional>
#include <msclr\marshal.h>
#include <msclr\marshal_cppstd.h>
#include <msclr\marshal_windows.h>
#include "GuiProvider.h"

using namespace System;

ItemChangedCallbackType gCallBackToUnManaged = nullptr;

static void CallBackFromManaged(GuiProxy::GuiProvider::ItemChangedEventArgs item)
{
	using namespace msclr::interop;
	System::String^ k = item.key;
	System::String^ v = item.value;
	int m = (int)item.mode;
	

	ItemChangedArgs args;

	msclr::interop::marshal_context context;
	args.key = context.marshal_as<const wchar_t*>(k);
	args.val = context.marshal_as<const wchar_t*>(v);
	args.changedMode = static_cast<ItemChangedMode>(m);
	args.type = context.marshal_as<const wchar_t*>(item.type);
	
	gCallBackToUnManaged(&args);
}

void netsettings_GuiProviderTest()
{

	int k = 0;
}



void netsettings_Create(GuiCreateParams* createParams)
{
	
	GuiProxy::GuiProvider::CreateParams^ managedCreateParams = gcnew GuiProxy::GuiProvider::CreateParams();
	managedCreateParams->templateFilePath = gcnew String(createParams->templateFilePath);
	managedCreateParams->userSettingsFilePath = gcnew String(createParams->userSettingsFilePath);
	managedCreateParams->callback = gcnew  GuiProxy::GuiProvider::OnItemChangedDelegate(CallBackFromManaged);

	gCallBackToUnManaged = createParams->callback;

	GuiProxy::GuiProvider::Initialize(managedCreateParams);
}

void netsettings_SetVisible(bool visible)
{
	GuiProxy::GuiProvider::SetVisible(visible);
}

File: GuiProxy/GuiProvider.cs
using NetSettings.Data;
using NetSettings.Forms;
using System.Collections.Generic;
using System.IO;

namespace GuiProxy
{
    public class GuiProvider
    {

        public struct ItemChangedEventArgs
        {
            public string key;
            public string value;
            public string type;
            public NetSettings.Data.ItemChangedMode mode;
        };

        public delegate void OnItemChangedDelegate(ItemChangedEventArgs args);

        public class CreateParams
        {
            public string templateFilePath;
            public string userSettingsFilePath;
            public OnItemChangedDelegate callback;
        }

        static SettingsForm form;
        static Dictionary<string, object> fData;
        private static CreateParams fCreateParams;
        static OnItemChangedDelegate OnItemChanged;
        static DataProvider fDataProvider;

        public static void Initialize(CreateParams createParams)
        {
            fCreateParams = createParams;
            OnItemChanged = createParams.callback;
            fDataProvider = new DataProvider(ItemTree.FromFile(createParams.templateFilePath));
            if (File.Exists(fCreateParams.userSettingsFilePath))
                fData = UserDataSerializer.LoadFromFile(fCreateParams.userSettingsFilePath);
            else 
                fData = fDataProvider.GenerateDefaultOptionsSet();

            fDataProvider.ItemChanged += DataProvider_ItemChanged;
            fDataProvider.DataBinding = fData;
            if (form == null)
            {
                form = new SettingsForm(fDataProvider);
                form.OnSave += Form_OnSave;
            }
            
        }

        private static void SaveUserSettingsFile()
        {
            UserDataSerializer.SaveToFile(fDataProvider.RootTemplate, fData, fCreateParams.userSettingsFilePath);
        }

        private static void Form_OnSave()
        {
            SaveUserSettingsFile();
        }

        public static void SetVisible(bool visible)
        {
            //if form is already visible - bring to front
            if (form.Visible == true && visible == true)
            {
                form.BringToFront();
            }
            else
            {
                form.Visible = visible;
            }
        }
        public static void SaveUserSettings()
        {
            SaveUserSettingsFile();
        }

        private static void DataProvider_ItemChanged(NetSettings.Data.ItemChangedArgs changedArgs)
        {
            ItemChangedEventArgs args;
            args.key = changedArgs.Key;
            args.value = changedArgs.Val != null ? changedArgs.Val.ToString().ToLower() : "";
            args.mode = changedArgs.ChangedMode;
            args.type = changedArgs.type;
            OnItemChanged(args);
        }
    }
}
